File: PA44.java
//106. Construct Binary Tree from Inorder and Postorder Traversal
//Time complexity : O(n) // n is the number of nodes in tree
//Space Complexity: O(n) // n is the height of tree in worst case it is equal to no of nodes in tree


class Solution {
    int index=0;
    HashMap<Integer, Integer> hm;
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        hm= new HashMap<>();
        for(int i=0;i<inorder.length;i++){
            hm.put(inorder[i],i);
        }
        index=postorder.length-1;
        return helper(inorder, postorder, 0, inorder.length-1);
    }
    
    public TreeNode helper(int[] inorder, int[] postorder, int start, int end){
        
        if(start>end || index<0){
            return null;
        }
        
        int rval= postorder[index];
        index--;
        TreeNode root = new TreeNode(rval);     
        int m= hm.get(rval);
        if(start==end){
            return root;
        }
        root.right= helper(inorder, postorder, m+1, end);
        root.left= helper(inorder, postorder, start, m-1);
        
        
        
        return root;
        
    }
}

File: PA_44.py
#LC 106 - Construct Binary Tree from Inorder and Postorder Traversal
#Time Complexity - O(n)
#Space Complexity - O(n)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        self.index = len(postorder) - 1
        self.d = {}
        for i in range(len(inorder)):
            self.d[inorder[i]] = i
        
        def helper(inorder, postorder,start,end):
            #base case
            if start > end or self.index < 0:
                return None
            root = TreeNode(postorder[self.index])
            self.index = self.index - 1
            inIdx = self.d[root.val]
            root.right = helper(inorder,postorder,inIdx+1,end)
            root.left = helper(inorder,postorder,start,inIdx-1)
            return root
        return helper(inorder,postorder,0,len(inorder)-1)